<a href="https://colab.research.google.com/github/alialaki/Applied-AI-ML-for-HR/blob/main/Applied_AI_for_Human_Resources.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#The first Featured use case that we will solve in this course is predicting Employee Attrition.
**Employee attrition is one of the major concens for a HR organization. When employees leave, it has many side effects. There is a loss of organization and product-specific expertise, loss of productivity due to new hires taking time to onboard. Sometimes employees have a great relationships with customers, and it is hard to rebuild. Thre are also hiring costs assosiated. Empolyees leave due to various reasons. They incloud compensation, work satisfaction, performance, and issues with supervisors. the online world had maded it easy for outside recruiters to approach employees with better job offer.**

#How can AI help in this employee attrition?
First, we need to collect 360-degree data about employee's past and present. This include:
 compensation and

*   Tenure in the company
*   performance ratings

*   compensation
*   promotions

*   sentiment
*   relationships the employee has with their supervisor and peers also play a key part

360 reviews will help understand this. once the data is collected and associated by both past and present employees, it provides input to build an ML model to predict attrition.

**Predict employeel's tendency to leave**

**Take preventive action that is needed**



In [ ]:
#!pip show tensorflow

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import tensorflow as tf
import sys

Mounted at /content/drive


In [ ]:
attrition_data = pd.read_csv("/content/drive/MyDrive/employee_attrition.csv")

In [ ]:
attrition_data.head(10)

EmployeeID  TotalMonthsOfExp  TotalOrgsWorked  MonthsInOrg  \
0           1               110                4            9   
1           2               103                3           51   
2           3                41                4           16   
3           4                32                4           17   
4           5                80                3           16   
5           6               109                1           39   
6           7                28                2           21   
7           8                81                2           35   
8           9                23                1            9   
9          10                86                1           70   

   LastPayIncrementBand  AverageFeedback  LastPromotionYears  Attrition  
0                     5                4                   4          1  
1                     1                4                   2          0  
2                     5                4                   4          1  
3                     5                2                   3          0  
4                     3                4                   2          0  
5                     4                2                   4          1  
6                     5                2                   4          1  
7                     3                4                   1          0  
8                     3                3                   4          1  
9                     1                1                   4          1

In [ ]:
attrition_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   EmployeeID            1000 non-null   int64
 1   TotalMonthsOfExp      1000 non-null   int64
 2   TotalOrgsWorked       1000 non-null   int64
 3   MonthsInOrg           1000 non-null   int64
 4   LastPayIncrementBand  1000 non-null   int64
 5   AverageFeedback       1000 non-null   int64
 6   LastPromotionYears    1000 non-null   int64
 7   Attrition             1000 non-null   int64
dtypes: int64(8)
memory usage: 62.6 KB


In [ ]:
attrition_data.describe()

EmployeeID  TotalMonthsOfExp  TotalOrgsWorked  MonthsInOrg  \
count  1000.000000       1000.000000      1000.000000  1000.000000   
mean    500.500000         61.002000         3.056000    31.000000   
std     288.819436         34.818132         1.398148    26.783341   
min       1.000000          1.000000         1.000000     0.000000   
25%     250.750000         31.000000         2.000000     9.000000   
50%     500.500000         61.000000         3.000000    23.000000   
75%     750.250000         91.250000         4.000000    48.000000   
max    1000.000000        120.000000         5.000000   116.000000   

       LastPayIncrementBand  AverageFeedback  LastPromotionYears    Attrition  
count           1000.000000      1000.000000          1000.00000  1000.000000  
mean               3.063000         2.539000             2.45400     0.230000  
std                1.413162         1.143585             1.10413     0.421043  
min                1.000000         1.000000             1.00000     0.000000  
25%                2.000000         1.000000             1.00000     0.000000  
50%                3.000000         3.000000             2.00000     0.000000  
75%                4.000000         4.000000             3.00000     0.000000  
max                5.000000         4.000000             4.00000     1.000000

In [ ]:
print(f"Data Loaded:\n----------------\n {attrition_data.dtypes}")

Data Loaded:
----------------
 EmployeeID              int64
TotalMonthsOfExp        int64
TotalOrgsWorked         int64
MonthsInOrg             int64
LastPayIncrementBand    int64
AverageFeedback         int64
LastPromotionYears      int64
Attrition               int64
dtype: object


**In Classification, it's always a good idea to understand the relationship between the feature and the target variables especially which features variables have the most impact on the target variable. WE do so using correlation analysis. Here we do a correlation analysis on the target variable called [attrition] **

In [ ]:
#Correlation Analysis of target attribute
attrition_data.corr()['Attrition']

EmployeeID             -0.036630
TotalMonthsOfExp        0.019702
TotalOrgsWorked         0.008706
MonthsInOrg             0.012605
LastPayIncrementBand    0.108528
AverageFeedback        -0.008253
LastPromotionYears      0.765641
Attrition               1.000000
Name: Attrition, dtype: float64

**We see that the LastPromotionYears (0.765641) has a significant impact on attrition. Meaning that employees leave when they don't see enough career growth.**

#**Next, we provide data for machine learning.**
#1- We first covert the dataset into a NumPy array of type float.
#This is the preferred input formant for Keras
#2- we split the feature and target variables into X & Y.

In [ ]:
#Convert to Dataframe to numpy array
np_attrition = attrition_data.to_numpy().astype(float)

#Create X_train with the first 7 attributes
X_train = np_attrition[:,1:7]

# #Create Y_train with attrition atribute
Y_train = np_attrition[:,7]

#Convert Y_train to one-hot encoding
#Since the attrition is bolean datatype, it has two unique values
Y_train = tf.keras.utils.to_categorical(Y_train,2)
print(f"X_train shape: {X_train.shape}")
print(f"Y_train shape: {Y_train.shape}")


X_train shape: (1000, 6)
Y_train shape: (1000, 2)


#02.05. Build Attrition model with Keras

In [ ]:
from os import name
from tensorflow import keras
from tensorflow.keras import optimizers
from tensorflow.keras.regularizers import l2

#Setup hyperprameter for deep learning
EPOCHS = 100
BATCH_SIZE = 100
#we can see the result if verbose set to 1. if I set to 0 it'll hide the result
VERBOSE = 1
#two calsses
NB_CLASSES = 2
N_HIDDEN = 128
#This means 20% of the data will be used for validation while building the model
VALIDATION_SPLIT = 0.2

#Create a Keras model
model = tf.keras.models.Sequential()

#Add first hidden Dense Layer
model.add(keras.layers.Dense(N_HIDDEN,
                             input_shape=(6,),
                             name='Dense-Layer-1',
                             activation='relu'))
model.add(keras.layers.Dense(N_HIDDEN,
                             name='Dense-layer-2',
                             activation='relu'))

#Add a final layer with softmax
model.add(keras.layers.Dense(NB_CLASSES,
                             name='Final',
                             activation='softmax'))
#Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

#Fit parameters
model.fit(X_train,
          Y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=VERBOSE,
          validation_split=VALIDATION_SPLIT)

Epoch 1/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 2s 33ms/step - accuracy: 0.4626 - loss: 3.9415 - val_accuracy: 0.8200 - val_loss: 1.5295
Epoch 2/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.6684 - loss: 1.2723 - val_accuracy: 0.8200 - val_loss: 0.6830
Epoch 3/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7632 - loss: 0.9056 - val_accuracy: 0.3700 - val_loss: 0.7703
Epoch 4/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.5183 - loss: 0.7366 - val_accuracy: 0.8200 - val_loss: 0.6097
Epoch 5/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7905 - loss: 0.5826 - val_accuracy: 0.8200 - val_loss: 0.5330
Epoch 6/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7825 - loss: 0.5298 - val_accuracy: 0.8200 - val_loss: 0.4849
Epoch 7/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7659 - loss: 0.5459 - val_accuracy: 0.8300 - val_loss: 0.4712
Epoch 8/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8029 - loss: 0.5080 - val_accuracy: 0.8300 - val_loss:

#**O2.06. Predict Attrition with Keras**

In [ ]:
attrition_data.columns

Index(['EmployeeID', 'TotalMonthsOfExp', 'TotalOrgsWorked', 'MonthsInOrg',
       'LastPayIncrementBand', 'AverageFeedback', 'LastPromotionYears',
       'Attrition'],
      dtype='object')

In [ ]:
TotalMonthsOfExp=40
TotalOrgsWorked=4
MonthsInOrg=20
LastPayIncrementBand=5
AverageFeedback=4
LastPromotionYears=4

# Create a NumPy array from the input list
input_data = np.array([[TotalMonthsOfExp, TotalOrgsWorked, MonthsInOrg,
       LastPayIncrementBand, AverageFeedback, LastPromotionYears]])

# Use model.predict to get the predicted probabilities
prediction = model.predict(input_data)

# Get the predicted class by finding the index of the maximum probability
predicted_class = np.argmax(prediction, axis=1)

print("Will employee leave ? (0: No Attrition, 1: Attrition)?", predicted_class[0])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step
Will employee leave ? (0: No Attrition, 1: Attrition)? 1


In [ ]:
#Bulk prediction

#Create a Bulk Numpy array for prediction
input_bulk=np.array([[111, 5, 85, 3, 2, 2],
                     [31, 2, 15, 4, 1, 4],
                     [61, 4, 24, 1, 4, 3],
                     [77, 4, 35, 3, 1, 1],
                     [81, 5, 7, 1, 2, 3],
                     [113, 4, 112, 5, 4, 1],
                     [101, 2, 48, 5, 1, 4],
                     [45, 4, 22, 5, 3, 1],
                     [25, 2, 2, 2, 3, 2],
                     [97, 3, 15, 3, 2, 4]])

prediction_bulk = model.predict(input_bulk)

predicted_class_bulk = np.argmax(prediction_bulk, axis=1)

print("Will employee leave ? (0: No Attrition, 1: Attrition)?", predicted_class_bulk)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 53ms/step
Will employee leave ? (0: No Attrition, 1: Attrition)? [0 1 0 0 0 0 1 0 0 1]
